File: cpp/04_014.cpp
#include <string>
#include <memory>

class Computer {
private:
    std::string cpu;
    std::string ram;
    
    // Private constructor only accessible by Builder
    Computer(const Builder& builder);
    
public:
    class Builder {
    private:
        std::string cpu;
        std::string ram;
        
    public:
        Builder& setCPU(const std::string& c) {
            cpu = c;
            return *this;
        }
        
        Builder& setRAM(const std::string& r) {
            ram = r;
            return *this;
        }
        
        Computer build() {
            return Computer(*this);
        }
        
        friend class Computer;
    };
};

Computer::Computer(const Builder& builder) : 
    cpu(builder.cpu), ram(builder.ram) {
    // Build computer with specifications
}

File: data structures and algorithms/cs/04_014.cs
public class Computer
{
    private Computer(Builder builder)
    {
        // Build computer with specifications
    }

    public class Builder
    {
        private string cpu;
        private string ram;

        public Builder SetCPU(string cpu)
        {
            this.cpu = cpu;
            return this;
        }

        public Computer Build()
        {
            return new Computer(this);
        }
    }
}
